### ISSUE TITLE
Create Terraform configuration for AKS

### ISSUE THREAD

**Issue by hferentschik** [jenkins-x/terraform-azurerm-jx#1]
Create Terraform configuration for Jenkins X cluster on Azure.

See also https://github.com/jenkins-x/terraform-google-jx and https://github.com/jenkins-x/terraform-aws-eks-jx.

The equivalent resources should be created for Azure.

**Comment by hferentschik** [comment]
Trickest part might be the preparation needed for the Vault operator. There is some information here https://github.com/banzaicloud/bank-vaults/blob/master/operator/deploy/cr-azure.yaml. Maybe it might make sense to address Vault in a separate issue?

Creating cluster, buckets and workload identity equivalent should be straight forward by mirroring the repositories mentioned above. Let's try to align structurally as well in supported variables and output.

**Comment by hferentschik** [comment]
Let's try to clarify what we are trying to achieve. 

We want to get into a state where we can create an AKS based cluster and then be able to run `jx boot` against it. Initially, we will aim for local secrets (no Vault; Vault setup we'll address in issue #3 and will require changes to `jx` itself) and no external DNS (we will use the default nip.io for DNS resolution). 

To achieve this we will need a minimum set of resources which are described below. The structure of the terraform module should be the same as for terraform-google-jx and terraform-aws-eks-jx. There should be a `main.tf`, together with `variables.tf` and `outputs.tf`. `main.tf` defines the providers w/ versions and then calls the submodules _backup_,  _cluster_,  _dns_, _vault_ each of which deals with a specific aspect of the required infrastructure. As part of this issue, we are focusing on _cluster_ and the resources which should be created there.

Primarily we need an AKS cluster. We should be able to provision it with as little input as possible defaulting to sensible values were possible. In the case of GKE the minimal required configuration is just the Google Cloud project id. Everything else gets defaulted. There are many configuration options for the cluster, but initially we expose for the user _cluster_name_, _node_machine_type_, _min_node_count_, _max_node_count_ (assuming these things are configurable with AKS). If no cluster name is specified we default to tf-<adjective>-<animal-name>. See https://github.com/jenkins-x/terraform-google-jx/blob/master/main.tf#L59.
It probably is also a good id to create a random id at the top-level Terraform file and pass this id into the submodules. Wherever resource names need a unique identifier to make names unique we can use this generated id. See https://github.com/jenkins-x/terraform-google-jx/blob/master/main.tf#L55.

Next, we need several storage buckets. Creation of these buckets should be conditional based on the variables _enable_log_storage_, _enable_report_storage_ and _enable_repository_storage_. The names of the buckets should be `logs-${var.cluster_name}-${var.cluster_id}`, `reports-${var.cluster_name}-${var.cluster_id}` and repository-${var.cluster_name}-${var.cluster_id}. See https://github.com/jenkins-x/terraform-google-jx/blob/master/modules/cluster/storage.tf

We also need a set of Kubernetes service accounts - kaniko_sa, tekton_sa, build_controller_sa. The idea is that these service accounts will use Managed Identity configuration to restrict which Cloud API resources they can create/modify. Which exact permissions each service account we need to investigate. Once the infrastructure is in place we will need to experiment with various permissions. At the very least all of these three service accounts need to be able to access the bucket API. They need to be able to read/write data into the created buckets (see above). In a first step we can give unrestricted access to the service accounts and address Managed Identity in #4.

**Comment by hferentschik** [comment]
For testing the created infrastructure we are using [shellspec](https://github.com/shellspec/shellspec) in _terraform-google-jx_ and _terraform-aws-eks-jx_. Preferably we do the same in terraform-azurerm-jx. The aim is to verify that the buckets exist, that the Kubernetes service accounts exist and they are limited in their permissions. How to do that is provider specific. In GKE we create a test Pod using a given service account. We then run a command inside the Pod (`gcloud auth list`) to ensure that Workload Identity is properly setup - https://github.com/jenkins-x/terraform-google-jx/blob/master/spec/k8s_spec.sh#L56. Not sure how to something equivalent for Azure.

**Comment by hferentschik** [comment]
It should always be possible to do a clean `terraform destroy`. Often created buckets cannot be deleted once they contain content. We use a _force_destroy_ variable to allow for full destruction of resources. The default is false to avoid accidental data loss, but it can be set to true for test or development cluster allowing a clean destruction of the cluster via Terraform.

**Comment by helayoty** [comment]
@deanesmith I can see the implementation for this issue has been merged. can we consider it as closed? any work still needed here?

**Comment by deanesmith** [comment]
Sounds good

**Comment by chrismellard** [comment]
We have a basic terraform module now so closing this issue
